src/radar_control.py

```diff
@@ -35,193 +36,258 @@
+        message.SOB = SOB
+        message.EOB = EOB
+        message.sequence_num = driver_params.seqnum_start + driver_params.num_sequences
+        message.numberofreceivesamples = driver_params.sequence.numberofreceivesamples
+        message.seqtime = driver_params.sequence.seqtime
+        message.align_sequences = driver_params.sequence.align_sequences
+
+        if pulse_transmit_data["isarepeat"]:
+            # antennas empty
+            # samples empty
+            # ctrfreq empty
+            # rxrate and txrate empty
+            log.debug("repeat timing", timing=timing, sob=SOB, eob=EOB)
+        else:
+            # Setup data to send to driver for transmit.
+            log.debug("non-repeat timing", timing=timing, sob=SOB, eob=EOB)
+            samples_array = pulse_transmit_data["samples_array"]
+            for ant_idx in range(samples_array.shape[0]):
+                sample_add = message.channel_samples.add()
+                # Add one Samples message for each channel possible in config.
+                # Any unused channels will be sent zeros.
+                # Protobuf expects types: int, long, or float, will reject numpy types and
+                # throw a TypeError so we must convert the numpy arrays to lists
+                sample_add.real.extend(samples_array[ant_idx, :].real.tolist())
+                sample_add.imag.extend(samples_array[ant_idx, :].imag.tolist())
+
+            message.txcenterfreq = (
+                driver_params.sequence.txctrfreq * 1000
+            )  # convert to Hz
+            message.rxcenterfreq = (
+                driver_params.sequence.rxctrfreq * 1000
+            )  # convert to Hz
+            message.txrate = driver_params.experiment.txrate
+            message.rxrate = driver_params.experiment.rxrate
+            message.numberofreceivesamples = (
+                driver_params.sequence.numberofreceivesamples
+            )
 
+    return message.SerializeToString()
 
-def send_dsp_metadata(
-    radctrl_to_dsp,
-    dsp_radctrl_iden,
-    radctrl_to_brian,
-    brian_radctrl_iden,
-    rxrate,
-    output_sample_rate,
-    seqnum,
-    slice_ids,
-    slice_dict,
-    beam_dict,
-    sequence_time,
-    first_rx_sample_start,
-    rxctrfreq,
-    pulse_phase_offsets,
-    decimation_scheme=None,
-):
+
+def dsp_comms_thread(radctrl_dsp_iden, dsp_socket_iden, router_addr):
+    """
+    Thread for handling communication between radar_control and rx_signal_processing.
+    """
+
+    inproc_socket = zmq.Context().instance().socket(zmq.PAIR)
+    inproc_socket.connect("inproc://radctrl")
+
+    radctrl_dsp_socket = socket_operations.create_sockets(
+        [radctrl_dsp_iden], router_addr
+    )[0]
+
+    while True:
+        message = inproc_socket.recv_pyobj()
+        # Wait for metadata from the main thread
+
+        bytes_packet = pickle.dumps(message, protocol=pickle.HIGHEST_PROTOCOL)
+        socket_operations.send_obj(radctrl_dsp_socket, dsp_socket_iden, bytes_packet)
```

Comment:
Added `send_pyobj` and `recv_pyobj` functions to `socket_operations.py` to act as wrappers to the `bytes` functions. The `send` function pickles the object it receives then sends it. The `recv` option un-pickles the object and optionally can check to ensure the received object is of the correct type.

All instances of pickling then sending or receiving bytes have been replaced except where `recv_bytes_from_any_iden` was used. This can be addressed in later changes if desired.